pkg/tidb/proxy.go

```diff
@@ -0,0 +1,226 @@
+package tidb
+
+import (
+	"io"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/pingcap/log"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+)
+
+type remote struct {
+	addr     string
+	inactive *atomic.Bool
+}
+
+func (r *remote) isActive() bool {
+	return !r.inactive.Load()
+}
+
+func (r *remote) becomeInactive() {
+	r.inactive.Store(true)
+	log.Debug("remote become inactive", zap.String("remote", r.addr))
+}
+
+func (r *remote) dial(timeout time.Duration) error {
+	conn, err := net.DialTimeout("tcp", r.addr, timeout)
+	if err != nil {
+		return err
+	}
+	conn.Close()
+	r.inactive.Store(false)
+	return nil
+}
+
+type proxy struct {
+	listener      net.Listener
+	checkInterval time.Duration
+	dialTimeout   time.Duration
+	doneCh        chan struct{}
+
+	// The key is server ddl id
+	remotes sync.Map
```

Comment:
I think it's better to not couple the proxy logic with any TiDB logic. Otherwise this proxy cannot be reused when we want to proxy more things. Can we simply use IpAddr as key?